File: dynamixel_workbench_practice/scripts/master_slave_joint_traj.py
#!/usr/bin/env python
import rospy

from sensor_msgs.msg import JointState
from trajectory_msgs.msg import JointTrajectory
from trajectory_msgs.msg import JointTrajectoryPoint
from dynamixel_workbench_msgs.srv import DynamixelCommand
from dynamixel_workbench_msgs.srv import DynamixelCommandResponse

# constant
control_cycle = 100

# shared msg data
traj = JointTrajectory()

# subscribe joint states and call dynamixel command service.
def joint_states_callback(msg):

    position = 2 * msg.position[0] - msg.position[1]
    # print "%s" % position
    # print "pan: %s" % msg.position[0]
    # print "tilt: %s" % msg.position[1]

    traj.header.seq = msg.header.seq
    traj.header.stamp = rospy.Time.now()
    traj.header.frame_id = msg.header.frame_id

    traj.joint_names = range(1)
    traj.joint_names = ['tilt']

    point = JointTrajectoryPoint()
    point.positions = range(1)
    point.positions = [position]
    # point.time_from_start = rospy.Duration.from_sec(0.008)
    point.time_from_start = rospy.Duration.from_sec(0.010)

    traj.points = range(1)
    traj.points = [point]

def joint_traj_handler():
    # initialize node
    rospy.init_node('dynamixel_command_client')

    # turn off torque of master (id=1)
    rospy.wait_for_service('dynamixel_workbench/dynamixel_command')
    try:
        service = rospy.ServiceProxy('/dynamixel_workbench/dynamixel_command', DynamixelCommand)
        response = service('', 1, 'Torque_Enable', 0)
        print response.comm_result
    except rospy.ServiceException, e:
        print "Service call failed: %s" % e
        return DynamixelCommandResponse

    # subscriber
    joint_states_sub = rospy.Subscriber('/dynamixel_workbench/joint_states', JointState, joint_states_callback, queue_size=1)

    # publisher
    joint_traj_pub = rospy.Publisher('/dynamixel_workbench/joint_trajectory', JointTrajectory, queue_size=1)

    # ROS loop w/ rate
    rate = rospy.Rate(control_cycle)
    while not rospy.is_shutdown():
        # print "loop"
        joint_traj_pub.publish(traj)
        rate.sleep()

if __name__ == "__main__":
    try:
        joint_traj_handler()
    except rospy.ROSInterruptException: pass

File: dynamixel_workbench_practice/src/master_slave_joint_traj.cpp
#include "dynamixel_workbench_practice/master_slave_joint_traj.h"

void jointStateCallback(const sensor_msgs::JointStateConstPtr& msg)
{
  double position = 2 * msg->position[0] - msg->position[1];

  traj.header.seq = msg->header.seq;
  traj.header.stamp = ros::Time::now();
  traj.header.frame_id = msg->header.frame_id;

  traj.joint_names.resize(1);
  // traj.joint_names.push_back("tilt");
  traj.joint_names[0] = "tilt";

  trajectory_msgs::JointTrajectoryPoint point;

  point.positions.resize(1);
  // point.positions.push_back(position);
  point.positions[0] = position;
  point.time_from_start = ros::Duration(1.0 / control_cycle);

  traj.points.resize(1);
  traj.points[0] = point;

  ROS_INFO("master : %f, slave : %f", msg->position[0], msg->position[1]);
  ROS_INFO("position : %f", position);

  ros::Rate loop_rate(control_cycle);
  loop_rate.sleep();

}

int main(int argc, char **argv)
{
  ros::init(argc, argv, "dynamixel_workbench_master_slave_demo");

  ros::NodeHandle nh;
  client = nh.serviceClient<dynamixel_workbench_msgs::DynamixelCommand>("/dynamixel_workbench/dynamixel_command");

  srv.request.command = "";
  srv.request.id = 1;
  srv.request.addr_name = "Torque_Enable";
  srv.request.value = 0;

  if (client.call(srv))
    {
      ROS_INFO("True (=1) or False (=0) : %d", srv.response.comm_result);
    }
  else
    {
      ROS_ERROR("Failed to call service.");
      return 1;
    }

  // ros::spinOnce(); // wait for the service request above
  ros::Publisher joint_traj_pub = nh.advertise<trajectory_msgs::JointTrajectory>("/dynamixel_workbench/joint_trajectory", queue);
  ros::Subscriber joint_sub = nh.subscribe("/dynamixel_workbench/joint_states", queue, jointStateCallback);

  // wait
  // ros::spinOnce();

  // infinite loop w/ rate
  ros::Rate loop_rate(control_cycle);
  while (ros::ok())
    {
      joint_traj_pub.publish(traj);
      ros::spinOnce();
      loop_rate.sleep();
    }

  return 0;
}
